torch/lib/THCUNN/SpatialDepthwiseConvolution.cu

```diff
@@ -0,0 +1,191 @@
+// updateOutput, updateGradInput Kernels ported from Sergey Zagoruyko's pyinn, which itself was a
+// port from Caffe
+
+#include "THCUNN.h"
+#include "THCDeviceTensor.cuh"
+#include "THCDeviceTensorUtils.cuh"
+#include "THCNumerics.cuh"
+#include "THCReduceApplyUtils.cuh"
+#include "THCSortUtils.cuh"
+#include "THCTensorMathReduce.cuh"
+#include "SharedMem.cuh"
+#include "common.h"
+
+template <typename T, typename IndexType>
+__global__ void spatialDepthwiseConvolutionUpdateOutput(
+    const THCDeviceTensor<T, 4> input,
+    THCDeviceTensor<T, 4> output,
+    const THCDeviceTensor<T, 4> weight,
+    const THCDeviceTensor<T, 1> bias,
+    bool biasEnabled,
+    IndexType totalElements,
+    const int outputChannels,
+    const int depthwiseMultiplier,
+    const int inputWidth, const int inputHeight,
+    const int outputWidth, const int outputHeight,
+    const int kernelWidth, const int kernelHeight,
+    const int strideWidth, const int strideHeight,
+    const int padWidth, const int padHeight,
+    const int dilationWidth, const int dilationHeight)
+{
+  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < totalElements;
+       linearIndex += gridDim.x * blockDim.x) {
+
+    const int n = linearIndex / outputChannels / outputHeight / outputWidth;
+    const int c = (linearIndex / outputHeight / outputWidth) % outputChannels;
```

Comment:
replacing divs by muls here might help (linearIndex / (outputChannels*outputHeight*outputWidth). Or restructuring the code somewhat, so that you start dividing by outputWidth, and proceed to dividing intermediate results by outputHeight and outputChannels. Also, templatizing on depthwiseMultiplier might help? Same for other kernels.

Reply:
I can optimize for trying to replace as many `/` with `*`. However, I was looking to avoid a lot of templating for these functions unless it was particularly beneficial. I feel like they could really be benefited if we were able to do some sort of run-time compilation for things like the filter size.